Move from chosen library to select2 and remove deprecated asset enqueuing strategy

jom opened the pull request:
Fixes #1202 
Fixes #1358

See the issue for the reasons we're moving away from `chosen`.
#### Changes proposed in this Pull Request:

* Removes support for old asset enqueuing strategy deprecated in #1354.
* Deprecate the old `chosen` library. 
* Loads `select2` asset for category select and multiselect.

This will eventually also be used in the backend for background pulled select fields (#846).

#### Testing instructions:

* This will involve quite a bit of testing for compat with popular plugins and themes.
* Generally, make sure all frontend screens (all shortcodes) have the same formatting as `master` and the `<select />` fields get the fancy select2 treatment.

<!-- Add the following only if this is meant to be in changelog -->
#### Proposed changelog entry for your changes:
* Developer: Moved from chosen JS library to select2.

jom committed:
Add select2 deps and build script

jom committed:
Add select2 v4.0.5

jom committed:
Fix assess copy task

jom committed:
Migrate from chosen to select2

jom committed:
Include chosen's CSS for legacy support

jom committed:
Fix for RTL issue

spencerfinnell commented:
I think the ‘enable’ filters should stay. Use that value as the base value for the new ones maybe?

jom commented:
@spencerfinnell Enqueue chosen with the enable filter (`job_manager_chosen_enabled`)? I was considering that and don't mind it if it will prevent issues (especially with older versions) of your themes. 

I suppose `job_manager_chosen_enabled` has also been an alias for "use WPJM core's multiselect scripts on this page" as well. We could set it as the default for now and maybe add a deprecation notice.

jom committed:
Enqueue chosen when old filter is used.

jom committed:
Deprecate `JOB_MANAGER_TEST_NEW_ASSET_BEHAVIOR` constant

Themes and plugins now need to explicitly tell WPJM when the frontend style is needed.

jom committed:
Register (but don't enqueue) select2 for admin usage

Used in our plugins and soon in the job listing editor in WP admin

jom committed:
Trigger event when application details is entirely visible

This helps with select2 issue in plugins.

tripflex commented:
Please keep in in the loop regarding this, the switch from Chosen to Select2 will affect a lot of my plugin's feature (working on an update for this) including conditional logic and other things that are specifically tied to Chosen.JS

jom commented:
@tripflex This is slated for our mid-January release. We should have a beta though in the next week.

spencerfinnell commented:
Thanks @jom -- is there a timeline for this release? It's very impactful as @tripflex mentioned and a roadmap/ETA will be helpful.

alexsanford reviewed (approved):
This looks good to me. Just one brief question below.

I did some quick testing on the shortcode pages and things seemed to work fine with select2. We'll do more comprehensive testing in beta, so I don't have any blockers 🙂

spencerfinnell commented:
@jom thanks. Just styles mainly. But the milestone timeline works for me. We can get a compatibility release out by then.

jom commented on Gruntfile.js:
```diff
@@ -277,6 +287,8 @@ module.exports = function( grunt ) {
 	grunt.loadNpmTasks( 'grunt-wp-readme-to-markdown');
 	grunt.loadNpmTasks( 'grunt-zip' );
 
+	grunt.registerTask( 'update-assets', [ 'copy:select2' ] );
```
It does need to be called manually. It will probably happen when we are prepping the release PR (not packaging/deploying it as it will be committed). I'll add it to our docs when we merge.

jom commented:
@donnapep I considered that, but (our plugins included) will break if we change the library because we call it in our non-core plugin scripts.

spencerfinnell commented:
> Drive-by comment, but am thinking it would be better to name the filter something that is not tied to the JS library name for better future-proofing.

Which is why theres a lot of side-stepping now. I'd say do what is needed for backwards compat but now but moving forward call it just something like `_enhanced_select_enabled`

jom committed:
Moved to generic filter for enhanced select

jom commented:
@spencerfinnell @donnapep I imagined it as a promise that we'll be including `select2` for WPJM related themes/plugins to use (without an abstraction layer in place) so they didn't need to package it as well. We would then be able to trigger a deprecation notice (like we are for `chosen`). For now, I've added d7ba2ba4659347aaa12d940d2574232d87f59dfd. 

I've updated the PRs for Resumes and Applications. In those implementations, until we package `select2` with them, I've tried to add checks for the existence of `$.fn.select2` before using it.

jom committed:
Move where select2 is registered in admin
